File: 1.14.c
#include <stdio.h>

/*
 *	print histogram of character frequency in iput
 *
 *	TODO:
 *		- print vertically
 *		- print for individual characters?
 */

int main()
{
	int c, i;
	int lengths[4];
	char types[4][6] = {"white", "digit", "alpha", "other"};
	float n, longest;

	for(i = 0; i < 4; ++i)
		lengths[i] = 0;

	while((c = getchar()) != EOF)
		if(c == ' ' || c == '\n' || c == '\t')
			lengths[0]++;
		else if(c >= '0' && c <= '9')
			lengths[1]++;
		else if(c >= 'A' && c <= 'z')
			lengths[2]++;
		else
			lengths[3]++;

	for(i = 1, longest = 0; i < 4;  ++i)
		if(lengths[i] > longest)
			longest = lengths[i];

	for(i = 0; i < 4; ++i)
	{
		printf("%s ", types[i]);
		for(n = 0; n < lengths[i]; n = n + (longest*longest) / (25*lengths[i]))
			printf("-");
		printf("\n");
	}
}

File: 1.14.py
#!/usr/bin/env python3

# print histogram of character types in input
# TODO: print vertically

from fileinput import input
from sys import stdout
from numpy import zeros

freqs = zeros(4)
types = ("white", "alpha", "digit", "other")

def histogram( names, freqs ) :
    most_freq = 0
    i = 0
    while i < len(freqs):
        if freqs[i] > most_freq:
            most_freq = freqs[i]
        i += 1
    i = 0
    while i < len(names):
        stdout.write(names[i])
        stdout.write(' ')
        n = 0
        while n < freqs[i]:
            stdout.write('-')
            n += pow(most_freq, 2) / (25 * freqs[i])
        stdout.write('\n')
        i += 1
    return


for line in input():
    for c in line:
        if c in (' ', '\t', '\n'):
            freqs[0] += 1
        elif c.isalpha():
            freqs[1] += 1
        elif c.isalnum():
            freqs[2] += 1
        else:
            freqs[3] += 1
histogram(types, freqs)
